src/postgres-util/src/lib.rs

```diff
@@ -207,15 +206,139 @@ pub async fn drop_replication_slots(config: Config, slots: &[&str]) -> Result<()
+                task::spawn(|| connection_task_name, connection);
+                Ok(client)
+            }
+            SshTunnelConfig::Tunnel { .. } => {
+                let tls = MakeTlsConnect::<TokioTcpStream>::make_tls_connect(&mut tls, &self.host)?;
+                let channel = self.establish_ssh_connection().await?;
+                let (client, connection) = self.postgres_config.connect_raw(channel, tls).await?;
+                task::spawn(|| connection_task_name, connection);
+                Ok(client)
+            }
+        }
+    }
+
+    /// Starts a replication connection to the upstream database
+    pub async fn connect_replication(mut self) -> Result<Client, anyhow::Error> {
+        let mut tls = make_tls(&self.postgres_config)?;
+        match self.ssh_tunnel {
+            SshTunnelConfig::Direct => {
+                let (client, connection) = self
+                    .postgres_config
+                    .replication_mode(ReplicationMode::Logical)
+                    .connect_timeout(Duration::from_secs(30))
+                    .keepalives_idle(Duration::from_secs(10 * 60))
+                    .connect(tls)
+                    .await?;
+                task::spawn(|| "postgres_connect_replication", connection);
+                Ok(client)
+            }
+            SshTunnelConfig::Tunnel { .. } => {
+                let tls = MakeTlsConnect::<TokioTcpStream>::make_tls_connect(&mut tls, &self.host)?;
+                let channel = self.establish_ssh_connection().await?;
+                let (client, connection) = self
+                    .postgres_config
+                    .replication_mode(ReplicationMode::Logical)
+                    .connect_timeout(Duration::from_secs(30))
+                    .keepalives_idle(Duration::from_secs(10 * 60))
+                    .connect_raw(channel, tls)
+                    .await?;
+                task::spawn(|| "postgres_connect_replication", connection);
+                Ok(client)
+            }
+        }
+    }
+
+    async fn establish_ssh_connection(
+        &self,
+    ) -> Result<AsyncChannel<TokioTcpStream>, anyhow::Error> {
+        match &self.ssh_tunnel {
+            SshTunnelConfig::Tunnel {
+                host,
+                port,
+                user,
+                public_key,
+                private_key,
+            } => {
+                let tcp_stream = TokioTcpStream::connect((host.as_str(), *port)).await?;
+                let mut session = AsyncSession::new(tcp_stream, Some(SessionConfiguration::new()))?;
+                session.handshake().await?;
+
+                session
+                    .userauth_pubkey_memory(user, Some(public_key), private_key, None)
+                    .await?;
+
+                if !session.authenticated() {
+                    bail!("failed SSH authentication")
+                };
+
+                // Advertise the source connection to Postgres as coming from the bastion host
```

Comment:
til wut this `host` field does

Reply:
It is actually important, because it's what the Postgres HBA is going to see! I'm not sure if defaulting to identifying as coming from the bastion host is ideal here, but it seemed as a sane default in case customers restrict via HBA to only accept connections from their local network.

(which is actually my testing setup at home, which is why I implemented it this way :p)